utils/waitable_counter.hh

```diff
@@ -0,0 +1,142 @@
+#include <seastar/core/gate.hh>
+#include <seastar/core/shared_future.hh>
+
+namespace utils {
+
+/**
+ * Simple counter type wrapper, that allows
+ * waiting for it to reach a given value.
+ * 
+ * Actual content must be monotonically 
+ * increasing, and unless you want a lot
+ * of entries, waiting should be either
+ * sparse, or at least quantized to "typical"
+ * values (like disk blocks or similar, hint hint)
+ */
+template<typename T, typename Cmp = std::less<T>>
+class waitable_counter {
+private:
+    using shared_promise_type = seastar::shared_promise<>;
+
+    T _value;
+    std::map<T, shared_promise_type, Cmp> _waiters;
+
+    using value_type = typename decltype(_waiters)::value_type;
+
+    template<typename Func>
+    waitable_counter& set(T v, Func&& f) {
+        if (_waiters.key_comp()(v, _value)) {
+            throw std::invalid_argument("Values must be monotonically increasing");
+        }
+        _value = std::move(v);
+        auto i = _waiters.upper_bound(_value);
+        std::for_each(_waiters.begin(), i, f);
+        _waiters.erase(_waiters.begin(), i);
+        return *this;
+    }
+
+public:
+    using clock = typename shared_promise_type::clock;
+    using time_point = typename shared_promise_type::time_point;
+
+    waitable_counter(T&& t = {}, Cmp cmp = {})
+        : _value(t)
+        , _waiters(std::move(cmp))
+    {}
+    /**
+     * Waits for the counter to reach pos
+     */
+    future<> wait(const T& pos, time_point timeout = time_point::max()) {
+        // if value held is already greater or 
+        // equal, just go ahead.
+        if (!_waiters.key_comp()(_value, pos)) {
+            return make_ready_future<>();
+        }
+        // else create/use the shared_promise
+        // at this pos.
+        auto& p = _waiters[pos];
+        return p.get_shared_future(timeout);
+    }
+    /**
+     * Sets the new counter value, notifying
+     * any waiting entities (for values <= v)
+     * 
+     * v must be >= current value.
+     */
+    waitable_counter& set(T v) {
+        return set(std::move(v), [](value_type& p) {
+            p.second.set_value();
+        });
+    }
+
+    /**
+     * Signals all waiters at v or below
+     * with exception. Current value is updated 
+     * to v, thus waits after this will pass
+     * as if successful.
+     */
+    waitable_counter& set_exception(T v, std::exception_ptr e) {
+        return set(std::move(v), [&](value_type& p) {
+            p.second.set_exception(e);
+        });
+    }
+
+    future<> wait_and_set_range(const T& s, T v) {
+        co_await wait(s);
+        set(std::move(v));
+    }
+
+    future<> wait_and_set_range_exception(const T& s, T v, std::exception_ptr e) {
+        co_await wait(s);
+        set_exception(std::move(v), std::move(e));
+    }
+
+    waitable_counter& operator=(T v) {
+        return set(std::move(v));
+    }
+
+    waitable_counter& operator+=(T v) {    
+        return set(std::move(v) + _value);
+    }
```

Comment:
Better to avoid IMO. These have serious side effects and should not be hidden behind an innocent interface such as +=.

Reply:
Otoh, they make it easier to replace the type in source code -> fewer changes down the line if need be. As long as the semantic meaning of the overloaded ops are sane, I don't really see a problem.

Reply:
"+=" typically means "add that to this". Not "add that to this, and trigger a bunch of other operations".

It's perfectly fine for someone who knows all the code by heart, terrible for someone who searches where things are triggered.

Reply:
Fair point, though it can be argued that adding to any counter whose value is "polled"/read/etc and causes side effects has the same properties, but sure.

Reply:
Fixed